unixfs/src/file/reader.rs

```diff
@@ -0,0 +1,333 @@
+    end: Ending,
+    had_links: bool,
+    links: Vec<PBLink<'a>>,
+    data: &'a [u8],
+    blocksizes: Vec<u64>,
+    metadata: FileMetadata,
+}
+
+impl AsRef<FileMetadata> for FileReader<'_> {
+    fn as_ref(&self) -> &FileMetadata {
+        &self.metadata
+    }
+}
+
+// TODO: this could be Range ... It just seemed there seems to be "two kinds" of endings but in
+// reality these are closer to two kinds of ranges or spans.
+#[derive(Debug)]
+enum Ending {
+    /// The block represented a subtree without actual content
+    TreeCoverage(u64),
+    /// The block repressented a leaf with actual content
+    Chunk(u64),
+}
+
+impl Ending {
+    /// Checks wheter or not the next range is good to be processed next.
+    fn check_is_suitable_next(&self, offset: u64, next: &Range<u64>) -> Result<(), FileError> {
+        match self {
+            Ending::TreeCoverage(cover_end) if next.start <= offset && &next.end > cover_end => {
+                // tree must be collapsing; we cant have root be some smaller *file* range than
+                // the child
+                Err(FileError::TreeExpandsOnLinks)
+            }
+            Ending::TreeCoverage(cover_end) if &next.start < cover_end && &next.end > cover_end => {
+                // when moving to sibling at the same height or above, it's coverage must start
+                // from where we stopped
+                //
+                // This has been separated instead of making the TreeExpandsOnLinks more general as
+                // this might be a reasonable way with unixfs to reuse lower trees but no such
+                // example has been found at least.
+                Err(FileError::TreeOverlapsBetweenLinks)
+            }
+            Ending::TreeCoverage(_) if next.start < offset => Err(FileError::EarlierLink),
+            Ending::Chunk(chunk_end) if &next.start != chunk_end => {
+                // when continuing on from leaf node to either tree at above or a chunk at
+                // next, the next must continue where we stopped
+                Err(FileError::TreeJumpsBetweenLinks)
+            }
+            _ => Ok(()),
+        }
+    }
+}
+
+impl<'a> FileReader<'a> {
+    /// Method for starting the file traversal. `data` is the raw data from unixfs block.
+    pub fn from_block(data: &'a [u8]) -> Result<Self, FileReadFailed> {
+        let inner = FlatUnixFs::try_from(data)?;
+        let metadata = FileMetadata::from(&inner.data);
+        Self::from_parts(inner, 0, metadata)
+    }
+
+    /// Called by Traversal to continue traversing a file tree traversal.
+    fn from_continued(
+        traversal: Traversal,
+        offset: u64,
+        data: &'a [u8],
+    ) -> Result<Self, FileReadFailed> {
+        let inner = FlatUnixFs::try_from(data)?;
+
+        if inner.data.mode.is_some() || inner.data.mtime.is_some() {
+            let metadata = FileMetadata::from(&inner.data);
+            return Err(FileError::NonRootDefinesMetadata(metadata))?;
```

Comment:
do we need the `?` at the end? the `return` should take care of early returning with the error, no?

Reply:
I am pretty sure 1.43 had much less of these, but of course I could had just fumbled.